File: macapp/Sources/SanitizerKit/Analyzer.swift
import Foundation

// The brain: turns a list of liked tracks into a reviewable Plan. Ported from
// analyzer.rb — the tunable knobs live in Options.
struct Analyzer {
    struct Options {
        var completionThreshold = 0.70
        var skitMaxSeconds = 60
        var dropUnplayable = true
        var findAlternatives = false
        var completeAlbums = true
    }

    // album_type preference for keeper selection: lower wins.
    static let albumTypeRank: [String: Int] = ["album": 0, "single": 1, "compilation": 2]

    var tracks: [Track]
    var library: LibraryProviding?
    var options: Options

    func buildPlan() async throws -> Plan {
        var plan = Plan()
        var kept = tracks

        if options.dropUnplayable { kept = try await dropUnplayable(kept, &plan) }
        kept = dedupe(kept, &plan)
        if options.completeAlbums, library != nil { try await completeAlbums(kept, &plan) }

        plan.stats = [
            "liked_tracks_scanned": tracks.count,
            "duplicates_removed":   plan.removals.filter { $0.reason.hasPrefix("duplicate") }.count,
            "unplayable_removed":   plan.removals.filter { $0.reason.hasPrefix("unplayable") }.count,
            "unplayable_replaced":  plan.replacements.count,
            "additions_suggested":  plan.additions.count,
            "albums_kept":          Set(kept.compactMap { $0.albumID }).count
        ]
        return plan
    }

    // MARK: - stages

    private func dropUnplayable(_ tracks: [Track], _ plan: inout Plan) async throws -> [Track] {
        let playable = tracks.filter { $0.playable }
        let dead = tracks.filter { !$0.playable }
        for track in dead {
            if let alt = try await alternative(for: track) {
                plan.replace(track, with: alt,
                             reason: "unplayable in your market — same recording (ISRC) plays here")
            } else {
                plan.remove(track, reason: "unplayable in your market")
            }
        }
        return playable
    }

    private func alternative(for track: Track) async throws -> Track? {
        guard options.findAlternatives, let library else { return nil }
        return try await library.findAlternative(track)
    }

    // Collapse "same recording, different release" clusters down to one keeper.
    private func dedupe(_ tracks: [Track], _ plan: inout Plan) -> [Track] {
        var kept: [Track] = []
        for (_, cluster) in Dictionary(grouping: tracks, by: { $0.fuzzyKey }) {
            guard cluster.count > 1 else { kept.append(cluster[0]); continue }
            let keeper = cluster.min { rankKey($0) < rankKey($1) }!
            kept.append(keeper)
            for loser in cluster where loser.id != keeper.id {
                plan.remove(loser, reason: duplicateReason(loser, keeper), keeper: keeper)
            }
        }
        return kept
    }

    // Ordered keeper rules: playable > explicit > album>single>compilation >
    // earliest added_at.
    private func rankKey(_ t: Track) -> (Int, Int, Int, String) {
        (t.playable ? 0 : 1,
         t.explicit ? 0 : 1,
         Analyzer.albumTypeRank[t.albumType ?? ""] ?? 9,
         t.addedAt ?? "")
    }

    private func duplicateReason(_ loser: Track, _ keeper: Track) -> String {
        if !loser.explicit && keeper.explicit {
            return "duplicate — clean version, keeping explicit"
        }
        let loserRank = Analyzer.albumTypeRank[loser.albumType ?? ""] ?? 9
        let keeperRank = Analyzer.albumTypeRank[keeper.albumType ?? ""] ?? 9
        if loserRank > keeperRank {
            return "duplicate — \(loser.albumType ?? "?") version, keeping \(keeper.albumType ?? "?")"
        }
        return "duplicate — keeping one copy"
    }

    // For albums you've mostly liked (ignoring skits), suggest the missing songs.
    private func completeAlbums(_ kept: [Track], _ plan: inout Plan) async throws {
        guard let library else { return }
        let albums = Dictionary(grouping: kept.filter { $0.albumType == "album" && $0.albumID != nil },
                                by: { $0.albumID! })

        for (albumID, liked) in albums {
            let total = liked[0].albumTotalTracks
            if total == 0 { continue }

            let likedReal = liked.filter { !$0.isSkit(maxSeconds: options.skitMaxSeconds) }
            // Rough gate before spending an API call on the full tracklist.
            if Double(likedReal.count) / Double(total) < options.completionThreshold { continue }

            let full = try await library.albumTracks(albumID, albumName: liked[0].albumName)
            let likedIDs = Set(liked.compactMap { $0.id })
            let missing = full
                .filter { track in !(track.id.map { likedIDs.contains($0) } ?? false) }
                .filter { !$0.isSkit(maxSeconds: options.skitMaxSeconds) }
            if missing.isEmpty { continue } // already complete (minus skits)

            let realTotal = full.filter { !$0.isSkit(maxSeconds: options.skitMaxSeconds) }.count
            if realTotal == 0 || Double(likedReal.count) / Double(realTotal) < options.completionThreshold { continue }

            for track in missing {
                plan.add(track, reason: "you like \(likedReal.count)/\(realTotal) of \"\(liked[0].albumName)\"")
            }
        }
    }
}

File: lib/spotify_sanitizer/analyzer.rb
# frozen_string_literal: true

module SpotifySanitizer
  # The brain. Turns a list of liked tracks into a reviewable Plan by applying
  # the dedup + curation rules. Pure logic apart from the optional album-tracklist
  # fetches used to make completion checks accurate.
  #
  # Tunable knobs live in DEFAULTS — these are the heuristics worth arguing about.
  class Analyzer
    DEFAULTS = {
      # Fraction of an album's non-skit tracks you must already like before we
      # suggest completing the rest.
      completion_threshold: 0.70,
      # Tracks at or under this length are treated as skits/interludes:
      # excluded from completion math and never proposed for addition.
      skit_max_seconds:     60,
      # Drop liked tracks that are unplayable in your market (greyed out).
      drop_unplayable:      true,
      # Suggest completing partially-liked albums.
      complete_albums:      true
    }.freeze

    # album_type preference for keeper selection: lower wins.
    ALBUM_TYPE_RANK = { "album" => 0, "single" => 1, "compilation" => 2 }.freeze

    def initialize(tracks, library: nil, options: {})
      @tracks  = tracks
      @library = library
      @opts    = DEFAULTS.merge(options)
    end

    def build_plan
      plan = Plan.new
      kept = @tracks

      kept = drop_unplayable(kept, plan) if @opts[:drop_unplayable]
      kept = dedupe(kept, plan)
      complete_albums(kept, plan) if @opts[:complete_albums] && @library

      plan.stats.merge!(
        liked_tracks_scanned: @tracks.size,
        duplicates_removed:   plan.removals.count { |r| r.reason.start_with?("duplicate") },
        unplayable_removed:   plan.removals.count { |r| r.reason.start_with?("unplayable") },
        additions_suggested:  plan.additions.size,
        albums_kept:          kept.map(&:album_id).compact.uniq.size
      )
      plan
    end

    private

    def drop_unplayable(tracks, plan)
      playable, dead = tracks.partition(&:playable)
      dead.each { |t| plan.remove(t, reason: "unplayable in your market") }
      playable
    end

    # Collapse "same recording, different release" clusters down to one keeper.
    def dedupe(tracks, plan)
      kept = []
      tracks.group_by(&:fuzzy_key).each_value do |cluster|
        if cluster.size == 1
          kept << cluster.first
          next
        end

        keeper = cluster.min { |a, b| compare_versions(a, b) }
        kept << keeper
        (cluster - [keeper]).each do |loser|
          plan.remove(loser, reason: duplicate_reason(loser, keeper), keeper: keeper)
        end
      end
      kept
    end

    # Returns negative if `a` is the better keeper. Ordered rules:
    #   1. playable beats unplayable
    #   2. explicit beats clean   (your rule)
    #   3. album beats single beats compilation
    #   4. stable tie-break: the one you liked earliest
    def compare_versions(a, b)
      rank.call(a) <=> rank.call(b)
    end

    def rank
      @rank ||= lambda do |t|
        [t.playable ? 0 : 1, t.explicit ? 0 : 1, ALBUM_TYPE_RANK.fetch(t.album_type, 9), t.added_at.to_s]
      end
    end

    def duplicate_reason(loser, keeper)
      if !loser.explicit && keeper.explicit
        "duplicate — clean version, keeping explicit"
      elsif ALBUM_TYPE_RANK.fetch(loser.album_type, 9) > ALBUM_TYPE_RANK.fetch(keeper.album_type, 9)
        "duplicate — #{loser.album_type} version, keeping #{keeper.album_type}"
      else
        "duplicate — keeping one copy"
      end
    end

    # For albums you've mostly liked (ignoring skits), suggest the missing songs.
    def complete_albums(kept, plan)
      by_album = kept.select { |t| t.album_type == "album" && t.album_id }
                     .group_by(&:album_id)

      by_album.each do |album_id, liked|
        total = liked.first.album_total_tracks.to_i
        next if total.zero?

        liked_real = liked.reject { |t| t.skit?(max_seconds: @opts[:skit_max_seconds]) }
        # Rough gate before spending an API call on the full tracklist.
        next if (liked_real.size.to_f / total) < @opts[:completion_threshold]

        full = @library.album_tracks(album_id)
        liked_ids = liked.map(&:id).to_set
        missing = full.reject { |t| liked_ids.include?(t.id) }
                      .reject { |t| t.skit?(max_seconds: @opts[:skit_max_seconds]) }
        next if missing.empty? # already complete (minus skits) — leave it alone

        # Only suggest if the real completion (non-skit) clears the threshold.
        real_total = full.reject { |t| t.skit?(max_seconds: @opts[:skit_max_seconds]) }.size
        next if real_total.zero? || (liked_real.size.to_f / real_total) < @opts[:completion_threshold]

        missing.each do |t|
          have = liked_real.size
          plan.add(t, reason: "you like #{have}/#{real_total} of \"#{liked.first.album_name}\"")
        end
      end
    end
  end
end
